src/jdk.jdwp.agent/unix/native/libjdwp/exec_md.c

```diff
@@ -44,6 +48,102 @@ static char *skipNonWhitespace(char *p) {
+// entry in "/proc/self/fd" (or its equivalent). Standard
+// input/output/error file descriptors will not be closed
+// by this function. This function returns 0 on failure
+// and 1 on success.
+int
+closeDescriptors(void)
+{
+    DIR *dp;
+    struct dirent *dirp;
+    /* leave out standard input/output/error descriptors */
+    int from_fd = STDERR_FILENO + 1;
+
+    /* We're trying to close all file descriptors, but opendir() might
+     * itself be implemented using a file descriptor, and we certainly
+     * don't want to close that while it's in use.  We assume that if
+     * opendir() is implemented using a file descriptor, then it uses
+     * the lowest numbered file descriptor, just like open().  So
+     * before calling opendir(), we close a couple explicitly, so that
+     * opendir() can then use these lowest numbered closed file
+     * descriptors afresh.   */
+
+    close(from_fd);          /* for possible use by opendir() */
+    close(from_fd + 1);      /* another one for good luck */
+    from_fd += 2; /* leave out the 2 we just closed, which the opendir() may use */
+
+#if defined(_AIX)
+    /* set FD_DIR for AIX which does not understand '/proc/self' - it
+     * requires the real process ID */
+    char aix_fd_dir[32];     /* the pid has at most 19 digits */
+    snprintf(aix_fd_dir, 32, "/proc/%d/fd", getpid());
+#endif
+
+    if ((dp = opendir(FD_DIR)) == NULL) {
+        LOG_MISC(("warning: failed to open dir %s while determining"
+                  " file descriptors to close for process %d", FD_DIR, getpid()));
+        return 0; // failure
+    }
+
+    while ((dirp = readdir(dp)) != NULL) {
+        int fd;
+        if (isdigit(dirp->d_name[0]) &&
+            (fd = strtol(dirp->d_name, NULL, 10)) >= from_fd) {
+            (void)close(fd);
+        }
+
+    }
+
+    (void)closedir(dp);
+
+    return 1; // success
+}
+
+// Does necessary housekeeping of a forked child process
+// (like closing copied file descriptors) before
+// execing the child process. This function never returns.
+void
+forkedChildProcess(const char *file, char *const argv[])
+{
+    /* Close all file descriptors that have been copied over
+     * from the parent process due to fork(). */
+    if (closeDescriptors() == 0) { /* failed,  close the old way */
+        /* Find max allowed file descriptors for a process
+         * and assume all were opened for the parent process and
+         * copied over to this child process. We close them all. */
+        const rlim_t max_fd = sysconf(_SC_OPEN_MAX);
+        JDI_ASSERT(max_fd != (rlim_t)-1);
+        /* leave out standard input/output/error file descriptors */
+        rlim_t i = STDERR_FILENO + 1;
+        LOG_MISC(("warning: failed to close file descriptors of"
+                  " child process optimally, falling back to closing"
+                  " %d file descriptors sequentially", (max_fd - i + 1)));
+        for (; i < max_fd; i++) {
+            (void)close(i);
```

Comment:
Here as well, I'd really like to check for `close()`for errors here and report any.

And if we find errors here, do we really want to proceed knowing that something went wrong already?